server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGateDetailsFormatter.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.qualitygate.ws;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.db.component.SnapshotDto;
+import org.sonarqube.ws.WsQualityGates.ProjectStatusWsResponse;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static org.sonar.api.utils.DateUtils.formatDateTime;
+
+public class QualityGateDetailsFormatter {
+  @CheckForNull
+  private final String measureData;
+  private final SnapshotDto snapshot;
+  private final ProjectStatusWsResponse.ProjectStatus.Builder projectStatusBuilder;
+
+  public QualityGateDetailsFormatter(@Nullable String measureData, SnapshotDto snapshot) {
+    this.measureData = measureData;
+    this.snapshot = snapshot;
+    this.projectStatusBuilder = ProjectStatusWsResponse.ProjectStatus.newBuilder();
+  }
+
+  public ProjectStatusWsResponse.ProjectStatus format() {
+    if (isNullOrEmpty(measureData)) {
+      return newUndefinedProjectStatusWsResponse();
+    }
+
+    JsonParser parser = new JsonParser();
+    JsonObject json = parser.parse(measureData).getAsJsonObject();
+
+    ProjectStatusWsResponse.Status qualityGateStatus = measureLevelToQualityGateStatus(json.get("level").getAsString());
+    projectStatusBuilder.setStatus(qualityGateStatus);
+
+    formatConditions(json.getAsJsonArray("conditions"));
+    formatPeriods();
+
+    return projectStatusBuilder.build();
+  }
+
+  private void formatPeriods() {
+    ProjectStatusWsResponse.Period.Builder periodBuilder = ProjectStatusWsResponse.Period.newBuilder();
+    for (int i = 1; i <= 5; i++) {
+      periodBuilder.clear();
+      boolean doesPeriodExist = false;
+
+      if (!isNullOrEmpty(snapshot.getPeriodMode(i))) {
+        doesPeriodExist = true;
+        periodBuilder.setIndex(i);
+        periodBuilder.setMode(snapshot.getPeriodMode(i));
+        if (snapshot.getPeriodDate(i) != null) {
+          periodBuilder.setDate(formatDateTime(snapshot.getPeriodDate(i)));
+        }
+        if (!isNullOrEmpty(snapshot.getPeriodModeParameter(i))) {
+          periodBuilder.setParameter(snapshot.getPeriodModeParameter(i));
+        }
+      }
+
+      if (doesPeriodExist) {
+        projectStatusBuilder.addPeriods(periodBuilder);
+      }
+    }
+  }
+
+  private void formatConditions(@Nullable JsonArray jsonConditions) {
+    if (jsonConditions == null) {
+      return;
+    }
+
+    for (JsonElement jsonCondition : jsonConditions) {
+      formatCondition(jsonCondition.getAsJsonObject());
+    }
+  }
+
+  private void formatCondition(JsonObject jsonCondition) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "formatCondition" is 15 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)